DM-4412 Development Workflow Document

jonathansick opened the pull request:
See the doc online: http://docs.lsst.codes/en/tickets-dm-4412/development/workflow.html

This doc is meant to be a one-stop resource for learning about how DM uses Git. The doc combines developer setup, Git best practices, DM policies, and documentation for Git LFS. In other words, it combines material that has previously been spread across many READMEs and Confluence pages.

I'd like JMatt to look at this from an LFS perspective. I'd like K-T to look at this from the perspective of documenting official policies and processes.

Some things to look for:

- I'm unsure of our official policy on PR titles and content (see commit messages).
- Does the organization work?

PaulPrice commented:
Please add suggestions on using [`__git_ps1`](https://dev.lsstcorp.org/trac/wiki/GitDemoAndTutorial#Someusefulcommand-lineprompthacking), which is a great aid to inexperienced and experienced git users alike.

I'm concerned that the LFS workflow section is duplicating information from afwdata and ci_hsc's README.md files.  Those should probably be changed to point here.

jonathansick commented:
@PaulPrice Good idea on adding git status to the command line as a best practice.

As for the README files, yes, that was the idea all along.

jonathansick commented:
From @jdswinbank 

> I'm not a reviewer and I don't know the prior history of this issue, so please take my comments with a pinch of salt. However: many of the existing project members, as well as those we hope to hire in future, are already experts in git (and a bunch of other stuff). What would be really helpful to me (and, I naively imagine, to them) is a simple list of LSST policies and rules. I don't want to have to wade through advice to use Homebrew, set up credential helpers, make a fancy prompt, and so on to get there. That's not to say that the more tutorial-style introduction for beginners isn't useful, but it shouldn't be the only way the core rules are presented.

A couple points here:

1. I totally get that we don't want to waste people's time with setup. *I don't need* the setup info, either. However, from experience we've already seen a lot of DM devs who don't have GitHub 2FA enabled, who don't have a credential helper, who don't know about Git LFS and how to install it. The idea of putting this sort of setup material in our process documentation is to ensure everyone has a functional setup. And though I included material on best practices for organizing git commits and writing commit messages, I'd say that we can all learn and get better at this. Having even this 'elementary' stuff is a reminder to all of us to be thoughtful and improve our craft.
2. Establishing that everyone has a functional credential helper makes it much easier to document things like Git LFS, for example.
2. To me it's weird to have the procedure in one place (https://confluence.lsstcorp.org/display/LDMDG/git+development+guidelines+for+LSST) and then the policy statement in another (https://confluence.lsstcorp.org/display/LDMDG/Git+Commit+Best+Practices and https://confluence.lsstcorp.org/display/LDMDG/DM+Branching+Policy). Why not just have the procedure checklist reflect and implement the official policy? That's the philosophy behind this doc.
3. You'll notice that I'm totally dropping all of the git tutorial content (e.g., https://confluence.lsstcorp.org/display/LDMDG/Git+Crash+Course + https://confluence.lsstcorp.org/display/LDMDG/Basic+Git+Operations + https://confluence.lsstcorp.org/display/LDMDG/Handling+Git+Push+Problems) since I don't think it's DM's job to teach git.

> While I'm here: it might also help to be clear about what the core rules actually are. Two things jump out at me:
The table of contents promises me "Suggestions for writing great commit messages" which links to "best practices" which contains a number of statements to the effect of "should XXX". Are those requirements? Suggestions? Something else?
What does "should" actually mean anyway? Is it a hard requirement? Are we assuming RFC 2119 or similar?

Yes, I'd like to have feedback from @ktlim on whether *all* docs with policy ramifications should be marked up [RFC 2119](http://rfc2119.openrfc.org) style. As I said in Point 2 above, my idea is that the procedures we write in docs should either be strong best practices, or reflect the actual policies. Rather than writing laws, my take is that we should just *show* people who the procedures work.

So let me summarize with a question. Do people like my approach of showing policy and best practice by example? Or should we go back to documenting laws and separating development tutorial material?

jdswinbank commented:
> So let me summarize with a question. Do people like my approach of showing policy and best practice by example? Or should we go back to documenting laws and separating development tutorial material?

Thanks for your thoughtful answer. It doesn't change my mind, though: I would like to see a clear specification of requirements, and then a separate set of hints/tips/getting started guidance. Of course, I'm under no illusion that this is a democracy.

(By the way, do we have a policy as to whether discussion like this belongs on PRs or on JIRA? I prefer the latter, since it seems like a more natural location to aggregate all the information about a ticket, but I note you've carefully transcribed my comments from there.)

jonathansick commented:
I thought @ktlim had decided all code review should be done on pull requests, but ironically, I'm not sure it's been written up. :)

jonathansick commented:
Thanks @jdswinbank. I need to go back and think a bit about an alternative outline (or someone can propose one) for a policy doc and a setup/tutorial doc. For example, a lot of my setup advice is to ensure that developers follow policy (e.g., make you GitHub address your institutional address; RFC-45).

jonathansick commented:
Question for the reviewers: would it be helpful if I pushed the [commit organization](http://docs.lsst.codes/en/tickets-dm-4412/development/git.html#commit-organization-best-practices) and [commit message](http://docs.lsst.codes/en/tickets-dm-4412/development/git.html#commit-message-best-practices) best practice docs down to the bottom as an appendix. Those are the least LSST policy specific of all the sections.

I can also move the 'Setting up Git' to the bottom as an appendix as well, and in its stead, have a bullet list of the minimum things you need to do to your git setup and don't provide any further help. E.g.

- have git > 1.8.2
- install git lfs and configure it (linking to the LFS section; merge credential helper docs into the LFS setup doc section)
- `git config --global user.email` and the GitHub primary email should be your institutional address

That way there'd be less heft at the top of the document, and bring the branch policy and review/merge policy docs closer to the top.

jonathansick commented:
So OpenStack is an interesting example of developer documentation. Like us, OpenStack devs are spread across many institutions, so the development is hugely decentralized and they need good docs to keep everyone working together. This is what their dev workflow docs look like: http://docs.openstack.org/infra/manual/developers.html

They take a fairly similar approach to the one I'm taking so far; by establishing policies through tutorials. At no point do they resort to RFC 2119 verbiage; it's simply understood that that their workflow is the de facto way to work on OpenStack.

Perhaps what I should do then is strip my current doc down to the essentials. I.e., if it's not something that we mandate (even if its considered best practice), then I just drop it from the doc or move it into an appendix.

jdswinbank commented:
> At no point do they resort to RFC 2119 verbiage; it's simply understood that that their workflow is the de facto way to work on OpenStack

To be clear: I think it's a problem to conflate suggestions with requirements. There are (at least) two ways you could avoid that problem: by using stilted RFC-like language to distinguish between them in a single text, or by separating them into two texts. I strongly prefer the latter, and was definitely not advocating writing our documentation as if it were an RFC!

The OpenStack documentation linked to above avoids takes the approach of making everything a requirement: it does not suggest or provide hints about how to perform tasks, it simply mandates a particular way of doing them. I don't much like this approach, but it avoids the potential for ambiguity.

jonathansick commented:
Okay, I'm going to do a re-org. Reviewers can hold on for a bit and I'll get new commits in :soon:

jonathansick commented:
Thanks to the review there some (hopefully good) scope re-alignment for this doc. The Git LFS part has been spun off into #5 and this doc is now an over Developer Workflow Policy Guide. I need to add words on JIRA usage policy and probably import words from a [trac doc on code reviews](https://dev.lsstcorp.org/trac/wiki/CodeReviewProcess).

[Let me know if the version on the web looks better](http://docs.lsst.codes/en/tickets-dm-4412/development/workflow.html), development will continue on this PR.

ktlim commented on development/workflow.rst:
```diff
@@ -0,0 +1,305 @@
+   git config --global user.name "Your Name"
+   git config --global user.email "your_email@institution.edu"
+
+Likewise, in your `GitHub account email settings <https://github.com/settings/emails>`_, add your institution-hosted email (e.g. @lsst.org).
```
"@lsst.org" is probably not the best example because apparently everyone gets such an alias, even if they're at another institution.  It's best to use the local E-mail (e.g. "@princeton.edu" or "@slac.stanford.edu").

ktlim commented on development/workflow.rst:
```diff
@@ -0,0 +1,305 @@
+
+``master`` is the main integration branch for our repositories.
+Development is not done directly on the ``master`` branch, but instead on *ticket branches*.
+*Trivial* documentation fixes are the only scenarios where working directly on ``master`` and by-passing the code review process is permitted.
```
Trivial fixes *or* additions to documentation can bypass code review (to encourage the latter); the documentation additions don't have to be trivial.

timj replied:
when we start doing doctests then this will get a bit more complicated.

PaulPrice replied:
The trivial fixes or additions to documentation that bypass code review are expected to have been tested on the user's machine.  Woe unto you if your direct-to-master commit breaks a test...

timj commented on development/workflow.rst:
```diff
@@ -0,0 +1,305 @@
+   git config --global user.email "your_email@institution.edu"
+
+Likewise, in your `GitHub account email settings <https://github.com/settings/emails>`_, add your institution-hosted email (e.g. @lsst.org).
+Set this institutional email as your **Primary GitHub** email address.
```
Why does it have to be primary? I might not want it to be my primary github address. Github is perfectly capable of associating addresses with different organizations so my personal email address is my primary address but all LSST org traffic goes to my LSST email account.

jonathansick replied:
My thinking what that if you make a hot fix to the docs on master using GitHub.com then you'll want that commit to still have the right email address. If we don't think anyone will do doc edits in the browser we can drop this.

ktlim commented on development/workflow.rst:
May want to mention "git add -p" as well to split commits into pieces.

timj commented on development/workflow.rst:
:+1:

ktlim commented on development/workflow.rst:
As long as you include the ticket number in the PR, the rest is whatever is useful to the reviewer.  But we shouldn't encourage putting a lot of context there, as it's then not visible in the code.

ktlim commented on development/workflow.rst:
These comments need to be in the "files changed" tab, not the "commits" tab, or they will get lost.

timj commented on development/workflow.rst:
```diff
@@ -0,0 +1,305 @@
+The master branch
+-----------------
+
+``master`` is the main integration branch for our repositories.
```
maybe note that we always expect master to be in a deployable state (at least in the sense that it passes unit and integration tests).

timj commented on development/workflow.rst:
It's interesting how Sims took a different road of `feature/SIM-NNN-short-summary-of-feature` for ticket branches. Maybe worthy of a footnote so that people don't get confused when they see repositories that are modified by Sims.

ktlim commented on development/workflow.rst:
The main goals are for the code to be maintainable by someone other than the author.  That means that the code must be readable, well-tested, and well-documented.
The code reviewer ideally should not have to be a design reviewer.  Any design review should have happened earlier in an RFC, an RFD, or other process (which we haven't really redefined after earlier design reviews were considered too heavyweight).

ktlim commented on development/workflow.rst:
```diff
@@ -0,0 +1,305 @@
+Development is not done directly on the ``master`` branch, but instead on *ticket branches*.
+*Trivial* documentation fixes are the only scenarios where working directly on ``master`` and by-passing the code review process is permitted.
+
+The Git history of ``master`` should never be re-written with force pushes.
```
I'd say "must" here.

PaulPrice commented on development/workflow.rst:
Please add something along the lines of "Review your own work with a discerning eye, as you would review someone else's work".

Do you want to mention how useful `git commit --fixup` is for fixing little things?

ktlim commented on development/workflow.rst:
If your home directory is on a network disk (NFS, AFS), you need to do a bit more.  See Git LFS docs.

ktlim commented on development/workflow.rst:
"change makes", but there are a lot of other typos that need cleanup that I'm not commenting on.

timj commented on development/workflow.rst:
or `git commit --amend`

ktlim commented on development/workflow.rst:
Except you can't safely comment on the Commits tab :(

ktlim commented on development/workflow.rst:
Or move those commits to the front.

ktlim commented on development/workflow.rst:
If you're working in a branch that has the ticket ID in its name, you don't need to add the ticket ID to the commit.

PaulPrice commented on development/workflow.rst:
Actually, I think there needs to be a separate section on rebasing (which can be referred to in preparing for review and responding to the review) because it's unfamiliar (and anathema) to everyone who's used to cvs and svn.  Then `git commit --fixup && git rebase -i` and `git commit --amend` can go in there.

PaulPrice commented on development/workflow.rst:
(So long as you `git merge --no-ff`.)

PaulPrice commented on development/workflow.rst:
I think you need something here about authority in the review.  Some of the audience will be scientists, who are used to an anonymous reviewer holding almost unlimited authority --- nothing gets by without the reviewer's say-so.  That's not how our reviews work.  On the other hand, I've seen software contributors blatantly ignore review comments and merge.  That's not how our reviews are supposed to work either.  I believe the idea is that the contributor and the reviewer (and anyone else who's interested) get together, discuss the merits of the contribution, and come to a consensus.  You don't merge until there's general agreement, but the reviewer isn't an anonymous tyrant, and you're always welcome to open an issue up to a larger audience (via Discourse or e-mail or RFC).

PaulPrice commented on development/workflow.rst:
PLEASE include more context than just the ticket number in the pull request title.  There are a lot of ticket numbers flying by at any one time, and those of us who haven't memorised them all need to have that context in the subject line of the e-mail notifications that github sends so we know if we should look closer.  I suggest putting the ticket number and the Jira title of the ticket.

jonathansick commented on development/workflow.rst:
So I've been doing this on my own work just so I can see what I'm doing in my repos without looking for the nearest merge commit. I can drop this recommendation though since it seems to be purely a personal preference.

jonathansick commented on development/workflow.rst:
:+1:

Do we like the text of https://dev.lsstcorp.org/trac/wiki/CodeReviewProcess ? I can incorporate that text into this section.

PaulPrice commented on development/workflow.rst:
Why is there a review process?

PaulPrice commented on development/workflow.rst:
I think it would be helpful to give a commit hash so that the reviewer can check he's got the version you meant for him to see (e.g., sometimes people forget to push).  To be perfectly clear, I usually dump the output of `git --no-pager log --stat --reverse origin/master..`, though maybe that's overkill.

PaulPrice commented on development/workflow.rst:
And so that the merge commit contains the ticket number.

PaulPrice commented on development/workflow.rst:
I've got a bunch of aliases I find helpful for our workflow.  Not sure you want to include them in this document as I'm sure there are matters of personal preference, but on the off chance it might be useful, here they are:

        stat = status
        co = checkout
        t = "tag -as"
        cp = cherry-pick
        desc = describe --tags --always
        recent = "log --graph --date-order --pretty=format:'%C(yellow)%h%d%Creset %Cred%an%Creset %Cblue%cr%Creset %s'"
        hist = "log --decorate --graph --date-order --stat"
        fixup = commit --fixup=HEAD
        # Amend last commit without modifying commit message
        amend = "!git log -n 1 --pretty=tformat:%s%n%n%b | git commit -F - --amend"
        # Outgoing commits
        out = "log @{u}.."
        # Undo merge
        unmerge = "reset --hard ORIG_HEAD"
        # Toss new commits
        toss = "reset --hard @{u}"
        # Clean up files that don't belong and toss any new commits
        nuke = "!git clean -df && git reset --hard @{u}"
        # Show new commits created by last command (e.g., after "git pull")
        new = !sh -c 'git log $1@{1}..$1@{0} "$@"'
        # List aliases
        alias = "config --get-regexp '^alias.*'"
        # Display for review
        review = !sh -c 'git log --patch --reverse "${2:-origin/master}".."${1:-HEAD}"' -
        # Display for submission
        sub = !sh -c 'git --no-pager log --stat --reverse "${2:-origin/master}".."${1:-HEAD}"' -
        sub-patch = !sh -c 'git --no-pager log --patch --reverse ${2:-origin/master}..${1:-HEAD}' -

PaulPrice commented on development/workflow.rst:
"Also did such-and-such" in the commit body is another good clue.

PaulPrice commented on development/workflow.rst:
The other side of the balancing act is that changes related to a single *feature* should be in the same commit, rather than spread over multiple commits.

PaulPrice commented on development/workflow.rst:
Set your `EDITOR` envvar.

PaulPrice commented on development/workflow.rst:
I find it useful to be explicit about what part of the codebase has been changed by prefixing the commit summary with an area or filename followed by a colon; I believe this is common on the Linux kernel and git.git.  Is this OK, encouraged, bad?

PaulPrice commented on development/workflow.rst:
If appropriate, what other approaches were considered for the problem and why they weren't used.

PaulPrice commented on development/workflow.rst:
I don't think there's anything in that link about authority in a review, and I've never seen any DM document or e-mail mention it.

ktlim commented on development/workflow.rst:
I don't really want to get too legalistic here.  It's a collaborative check-and-improve process.

ktlim commented on development/workflow.rst:
Cribbing text from the CodeReviewProcess would seem appropriate to answer that question.

jonathansick commented:
I think I've addressed the issues presented in the first round of comments, and created a new document that can serve as our workflow policy reference. Extraneous best practices are moved to appendices.

One suggestion was to include a rebasing tutorial; I'd like to save that for a later ticket since it would most likely be an appendix, and not related to policy specification.

ktlim commented on development/workflow.rst:
```diff
@@ -0,0 +1,667 @@
+   git config --global user.email "your_email@institution.edu"
+
+Likewise, in your `GitHub account email settings <https://github.com/settings/emails>`_, add your institution-hosted email.
+Set this institutional email as your **Primary GitHub** email address.
```
"We recommend that you set", rather than being *so* prescriptive?

ktlim commented on development/workflow.rst:
```diff
@@ -0,0 +1,667 @@
+   You can also assign tickets to others.
+   If you are uncertain about who the assignee should be you can allow the ticket to be automatically assigned (which defaults to the component's T/CAM; `RFC-51 <https://jira.lsstcorp.org/browse/RFC-51>`_).
+Description
+   For stories, you should specify the ticket's scope and outline your implementation design in this field.
```
Ideally there should be a clear description of the deliverable that can serve as a "definition of done".

ktlim commented on development/workflow.rst:
```diff
@@ -0,0 +1,667 @@
+   If the ticket is a story, you must specify what epic it belongs to with this field.
+   By definition, bug or enhancement-type tickets are not associated with an epic.
+Team
+   You must specify which DM team doing the work with this field, for accounting purposes.
```
"is doing"

ktlim commented on development/workflow.rst:
```diff
@@ -0,0 +1,667 @@
+
+Tickets are created with a status of 'Todo.'
+
+If a ticket was assigned to you by someone else, you can set the status to 'Acknowledged' once you agree to schedule the work.
```
Um, we don't have an "Acknowledged" status in the DM JIRA project.

jonathansick replied:
Indeed. I got this 'Acknowledged' workflow from the [JIRA at LSST User's Guide](https://confluence.lsstcorp.org/display/JIRAatLSSTUserGuide/Using+JIRA+on+LSST#UsingJIRAonLSST-IssueDetailScreenCustomization). I'll take it out.

ktlim commented on development/workflow.rst:
```diff
@@ -0,0 +1,667 @@
+If a ticket was assigned to you by someone else, you can set the status to 'Acknowledged' once you agree to schedule the work.
+
+Once a ticket is being actively worked on you can upgrade the ticket's status to 'In Progress.'
+
```
Do you want to describe "Won't Fix" and "Invalid" here?

jonathansick replied:
I've put in words for Won't Fix and Invalid in the new pushes.

ktlim commented on development/workflow.rst:
```diff
@@ -0,0 +1,667 @@
+In some circumstances, a ``release`` integration branch may be used by the release manager.
+Development is not done directly on the ``master`` branch, but instead on *ticket branches*.
+
+Documentation edits and additions are the only scenarios where working directly on ``master`` and by-passing the code review process is permitted.
```
Well, trivial (like one-character) fixes are allowed to be done that way as well, but maybe it's better not to mention it.

ktlim commented on development/workflow.rst:
Second "pull request" is redundant, I think.

ktlim commented on development/workflow.rst:
If there are multiple reviewers, the convention is for each reviewer to remove his/her name from the Reviewers list; the last one switches the status to "Reviewed".

ktlim commented on development/workflow.rst:
Missing trailing backquote.

ktlim commented on development/workflow.rst:
Is it 16 or 10?

jhoblitt commented on development/workflow.rst:
```diff
@@ -0,0 +1,668 @@
+
+Likewise, in your `GitHub account email settings <https://github.com/settings/emails>`_, add your institution-hosted email.
+We recommend that you set this institutional email as your **Primary GitHub** email address.
+This step ensures that Git commits you make `directly on GitHub.com <https://help.github.com/articles/github-flow-in-the-browser/>`_ (such as quick documentation fixes) have proper authorship metadata.
```
Perhaps note that github uses the primary email address as the authorship for PRs merged via the www UI.  This is currently problematic for me as I don't use @lsst.org as my primary GH email.

timj replied:
Problematic for me as well. I really wish Github would do the right thing. They already send all these messages to the correct email address based on the organization so it amazes me they can't use that same mapping for the commits.

jonathansick replied:
So solution is to just write a caveat that if you aren't willing to change your :octocat: email address then you shouldn't do in-browser commits and merges. I'm fine with that.

timj replied:
Since I'm wishing, I also wish that git could do per directory tree .git user/email configuration. Per-repo email configuration is unreliable (lsstsw trashes it) and `~/.gitconfig` is going to be wrong for 50% of my repositories...

jhoblitt commented on development/workflow.rst:
```diff
@@ -0,0 +1,668 @@
+Agile concepts
+--------------
+
+Issue
```
It would be nice to have link anchors for these sub-sub-subsections.  Is that possible?

jhoblitt commented on development/workflow.rst:
```diff
@@ -0,0 +1,668 @@
+
+Issue
+   Issues are the fundamental units of work/planning information in JIRA.
+Story Points
```
think a few more words on Story points would be useful, In particular spelling out that a week != 10 story points and that different velocities are assumed for different types of staff

jhoblitt commented on development/workflow.rst:
```diff
@@ -0,0 +1,668 @@
+   That is, stories are for work that accomplish your main goals for a cycle.
+Bug
+   A bug is an emergent (not planned with an epic) ticket that fixes a fault in the software that exists on ``master``.
+   Bug tickets are not associated with an epic.
```
This isn't completely true as they are typically assigned to an EPIC when SPs are claimed for resolving the bug. This may need more pedantic explanation.

jonathansick replied:
Can I get a few sentences from a T/CAM like @frossie on the matter of bugs/epics/earned value?

jhoblitt commented on development/workflow.rst:
```diff
@@ -0,0 +1,668 @@
+DM Git Branching Policy
+=======================
+
+Rather than forking LSST's GitHub repositories, DM developers use a *shared repository model* by cloning repositories in the `lsst <https://github.com/lsst>`_, `lsst-dm <https://github.com/lsst>`_, and `lsst-sqre <https://github.com/lsst>`_ GitHub organizations.
```
We generally use the fork & PR model for repos under `lsst-sqre`.  The shared model is only needed for repos built by `lsst_build`.

jonathansick replied:
I think it's understood that SQuaRE DevOps operates different rules than science pipelines. I don't want to have to make that complication for science pipelines developer docs.

jhoblitt commented on development/workflow.rst:
I think a screenshot would be helpful here.  Also a note about github auth against lsst orgs being required in order to run a build.

jhoblitt commented on development/workflow.rst:
Screenshot might be helpful

jhoblitt commented on development/workflow.rst:
Are we disallowing the GH merge button?

ktlim commented on development/workflow.rst:
We have not disallowed the Big Green Button (as long as the DM-NNNN makes it into the merge commit, which I believe it does as long as the other procedures are followed), but you pointed out the caveat that it requires the primary E-mail to be set correctly.  Is there a greater likelihood of creating "braided" history if we use the button vs. the commands above?

timj commented on development/workflow.rst:
Github doesn't tell you whether there have been any commits on master since the branch was made so you can't be certain you are going to end up with the proper testable history or not.

jonathansick commented on development/workflow.rst:
I've put the screenshot in.

jonathansick commented on development/workflow.rst:
Also put this screenshot in.

jonathansick commented:
I've updated the PR again. I think the main issues that still need to be resolved are:

1. Meaning of a 'bug' issue, and whether we actually assign them to epics to earn value (I'm a newbie so I don't know).
2. Whether we want to specifically ban in-browser commits if your email address is not correct.
3. Whether we want to specifically advocate the fork-and-pr model for repos outside of the stack build.

jonathansick commented:
@frossie has pointed out to me that no one uses the procedure described for multiple reviewers signing off on a ticket:

>  If there are multiple reviewers, our convention is that each review removes her/his name from the Reviews list to indicate sign-off; the final reviewer switches the status to ‘Reviewed.’ This indicates the ticket is ready to be merged.

Instead, the de facto meaning of having 'multiple reviewers' is that a reviewer is assigned until one actually does the review. How should we deal with the differing meanings of having multiple reviewers on a ticket?

timj commented:
I know that Qserv team do actually work like that (removing each reviewer as they review).

jonathansick commented:
I do agree that carpet bombing the reviewers list until something happens is probably not the best way to deal with getting a review done.

ktlim commented:
I think "no one uses" is an exaggeration, and I don't think that the "de facto" method is workable. My suggestion for dealing with differing meanings is to write down which meaning is acceptable, which I think your document does.

jonathansick commented:
Okay, better practices though documentation. :+1:
